README.md
<!--
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2019-11-04 09:28:37
 * @LastEditors: zhai
 * @LastEditTime: 2021-05-26 17:45:41
-->
# 燕姿鼓励师

在 VS Code 中连续写代码一小时（时间可配置），会有孙燕姿提醒你该休息啦~

## 使用

1. 除了每过一小时会自动弹出提醒页面，也可以按 `F1`, 然后输入 `syz: 打开提醒页面`来打开提醒页面
2. 在提醒页可以使用方向键/鼠标滚轮，循环、循环、再循环……翻来覆去看不同照片
3. 在花好约猿︱媛页面可以偶遇其他媛或者猿，一起吐槽下产品也不错~
4. 在状态栏可以打开/关闭提醒，以及设置提醒间隔时间
5. 如果五分钟没有操作vscode，那就认为已经在休息了。这时候会停止计时，只有下一次鼠标/键盘活动时才会重新计时
6. 欢迎来github拍砖


![img](https://cdn.jsdelivr.net/gh/kanbang/resource@main/pic/vsyanzi.gif)


## 配置

* `syz.reminderViewIntervalInMinutes`: 展示提醒页面的时间间隔（分钟）。(默认值为**60**)
* `syz.title`: 提示文字。 (默认值为**小哥哥，小哥哥，代码写久了，喝杯水吧~**)
* `syz.kedou`: true (启用蝌蚪聊聊)；false (停用蝌蚪聊聊)。(默认值为**true**)
* `syz.resLocal`: true (启用本地图片/视频)；false (停用本地图片/视频)。(默认值为**true**)
* `syz.resWeb`: true (启用网络图片/视频)；false (停用网络图片/视频)。(默认值为**true**)
* `syz.webResources`: 配置网络资源链接数组
* `syz.gender`: 设置性别 "男", "女", "保密" (默认值为**保密**)

```
如下例子，启用网络图片/视频，支持[jpg/png/mp4]格式：
"syz.resWeb": true,
"syz.webResources": [
    "https://github.com/kanbang/links/raw/master/pic/1.jpg",
    "https://github.com/kanbang/links/raw/master/pic/2.jpg",
    "https://github.com/kanbang/links/raw/master/video/Stefanie%20Sun/shenqi.mp4",
    "https://github.com/kanbang/links/raw/master/video/Stefanie%20Sun/banjuzaijian.mp4"
]
```
## 如何更改本地图片/视频

* vscode不允许读取外部文件路径，所以只能自己去替换插件内的媒体。替换步骤如下：
  
  1. 按 `F1`, 然后输入 `syzfolder: 打开媒体文件夹，添加自己喜欢的照片/视频` 来打开媒体文件夹

  2. 替换文件夹内的图片/视频


CHANGELOG.md
<!--
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2019-11-04 09:28:37
 * @LastEditors: zhai
 * @LastEditTime: 2021-05-26 17:54:35
-->


## 0.1.13 (2021-05-26)
* turn on/off reminder with statusbar button
* set reminder interval with statusbar button
* if not activity for five minutes, the timer will stop 
* update photos (*^_^*)

## 0.1.12 (2020-10-26)
* websocket which chat room used has been banned in webview. I'm trying to find another way...
* update photos (*^_^*)

## 0.1.11 (2020-9-5)
* update tadpole chat room
* switch photos by mouse click and wheel
* update photos (*^_^*)

## 0.1.9 (2020-8-28)
* remove mp4, cause current version of vscode electron does not include ffmpeg
* show random word from hitokoto-api
* update photos (*^_^*)

## 0.1.8 (2020-1-20)
* optimize text display 
* update photos (*^_^*)

## 0.1.7 (2019-11-21)
* adapt to chrome new autoplay-policy

## 0.1.6 (2019-10-17)
* update README.md
* little optimize

## 0.1.3 (2019-10-17)
* add nickname and gender configuration
* setting nickname and gender with gui

## 0.1.2 (2019-10-13)
* tadpole online 
* gender differentiation by color

## 0.1.1 (2019-08-6)
* add tadpole chat room for beta

## 0.1.0 (2019-07-30)
* optimize synchronization logic

## 0.0.10 (2019-07-26)
* fix bugs

## 0.0.9 (2019-07-24)
* not work in publish，so sad...

## 0.0.6 (2019-07-23)
* Happy Birthday to Stefanie Sun ^_^ 
* supports web video/picture resource links

## 0.0.5 (2019-07-04)
* vscode 1.36 has support ffmpeg, use your favorite video to remind ^_^ 

## 0.0.4 (2019-07-02)
* 按 `F1`, 然后输入 `syzfolder: 打开照片文件夹，添加自己喜欢的照片` 可以打开图片文件夹替换图片

## 0.0.3 (2019-06-25)
* bug fix
* try to support mp4, but failed,  because VS Code uses a version of electron that does not include ffmpeg. :sob: :sob: :sob:

## 0.0.2 (2019-06-25)
* more photos with slide by arrow key

## 0.0.1 (2019-06-24)
* Initial Release



src/extension.ts
/*
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2019-06-24 14:34:14
 * @LastEditors: zhai
 * @LastEditTime: 2021-05-26 16:35:12
 */
'use strict';

import * as vscode from 'vscode';

import Asset from './asset';
import { Reminder } from './util/reminder';
import { StatusBarItem } from 'vscode';
import { Utility } from './util/utility';
import * as path from 'path';
import * as dialog from './util/webviewDialog';
import * as fs from 'fs';

let reminder: Reminder;
let statusItemEnable: StatusBarItem;
let statusItemMinutes: StatusBarItem;
let asset: Asset;


const reminderOnText = "$(pass) 燕姿在线";
const reminderOffText = "$(circle-slash) 燕姿离线";
// smiley

export function activate(context: vscode.ExtensionContext) {
    asset = new Asset(context);

    let disposables = [

        vscode.commands.registerCommand('kk.syz.remind', () => {
            showStefanie(context);
        }),

        vscode.commands.registerCommand('kk.syz.chat', () => {
            showChat(context);
        }),

        vscode.commands.registerCommand('kk.syz.openPhotoFolder', () => {
            const path = asset.getLocalResPath();
            Utility.openFile(path);
        }),

        vscode.commands.registerCommand('kk.syz.switch_enable', () => {
            let config = vscode.workspace.getConfiguration("syz");
            let enable = config.get<boolean>('enable', true);

            if (enable) {
                config.update('enable', false);
                statusItemEnable.text = reminderOffText;
                statusItemMinutes.hide();
                vscode.window.showInformationMessage('syz reminder off');

                reminder.enable = false;
            }
            else {
                config.update('enable', true);
                statusItemEnable.text = reminderOnText;
                statusItemMinutes.show();
                vscode.window.showInformationMessage('syz reminder on');

                reminder.enable = true;
                reminder.minutespass = 0;
            }
        }),

        vscode.commands.registerCommand('kk.syz.set_minutes', () => {
            let config = vscode.workspace.getConfiguration("syz");

            let items: vscode.QuickPickItem[] = [
                { label: "1", description: "每[1]分钟提醒一次" },
                { label: "30", description: "每[30]分钟提醒一次" },
                { label: "45", description: "每[45]分钟提醒一次" },
                { label: "60", description: "每[60]分钟提醒一次" },
                { label: "90", description: "每[90]分钟提醒一次" },
                { label: "120", description: "每[120]分钟提醒一次" }
            ];

            vscode.window.showQuickPick(items).then(selection => {
                if (selection) {
                    statusItemMinutes.text = formatMinutesText(0, selection.label);
                    config.update('reminderIntervalInMinutes', parseInt(selection.label));

                    reminder.minutes = parseInt(selection.label);
                    reminder.minutespass = 0;
                }
            });
        })
    ];

    context.subscriptions.push(...disposables);

    try {
        // vscode.window.showInformationMessage("sync start...");
        asset.syncFiles();
    } catch (error) {
        // vscode.window.showErrorMessage("sync error...");
    }

    initReminder();
}


function formatMinutesText(minutespass: number, minutes: number | string) {
    return `$(clock) [${minutespass}/${minutes}]分钟`;
}

function initReminder() {

    statusItemEnable = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1100);
    statusItemEnable.command = 'kk.syz.switch_enable';
    statusItemEnable.show();

    statusItemMinutes = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1099);
    statusItemMinutes.command = 'kk.syz.set_minutes';

    let config = vscode.workspace.getConfiguration("syz");
    let enable = config.get<boolean>('enable', true);
    statusItemEnable.text = enable ? reminderOnText : reminderOffText;

    let minutes = config.get<number>('reminderIntervalInMinutes', 45);
    statusItemMinutes.text = formatMinutesText(0, minutes);

    if (enable) {
        statusItemMinutes.show();
    }

    reminder = new Reminder(
        async () => {
            let config = vscode.workspace.getConfiguration("syz");
            let kedou = config.get<boolean>('kedou', true);
            if (kedou) {
                await vscode.commands.executeCommand('kk.syz.chat');
            }

            await vscode.commands.executeCommand('kk.syz.remind');
        },
        (minutespass: number, minutes: number) => {
            statusItemMinutes.text = formatMinutesText(minutespass, minutes);
        }
    );

    reminder.minutes = minutes;
    reminder.enable = enable;

    vscode.workspace.onDidChangeTextDocument((ev) => {
		// console.log(ev);
		if (reminder) {
			reminder.active();
		}
	});

	vscode.window.onDidChangeTextEditorSelection((ev) => {
		// console.log(ev);
		if (reminder) {
			reminder.active();
		}
	});

	vscode.window.onDidChangeTextEditorVisibleRanges((ev) => {
		// console.log(ev);
		if (reminder) {
			reminder.active();
		}
	});
}


interface DialogResult {
    name: string;
}

var dlg_chat: dialog.WebviewDialog<DialogResult> | null = null;
var dlg_stefanie: dialog.WebviewDialog<DialogResult> | null = null;

async function showChat(context: vscode.ExtensionContext): Promise<void> {
    if (dlg_chat) {
        return;
    }

    // const rootDir = path.resolve(__dirname, '../kedou');
    const extPath = context.extensionPath;
    const rootDir = path.join(extPath, 'kedou');

    dlg_chat = new dialog.WebviewDialog<DialogResult>('webview-kedou', rootDir, '', 'index.html', vscode.ViewColumn.Beside,
        (message) => {
            switch (message.command) {
                case 'setting':
                    asset.setGender(message.gender);
                    asset.setNickname(message.nickname);
                    break;
                default:
                    break;
            }
        });

    dlg_chat.postMessage({
        command: 'setting',
        gender: asset.getGender(),
        nickname: asset.getNickname()
    });

    const result = await dlg_chat.getResult();
    dlg_chat = null;

    if (result) {
        // vscode.window.showInformationMessage("Webview dialog result: " + JSON.stringify(result));
    } else {
        // vscode.window.showInformationMessage("The webview dialog was cancelled.");
    }
}

async function showStefanie(context: vscode.ExtensionContext): Promise<void> {
    if (dlg_stefanie) {
        return;
    }

    const title = asset.getTitle();

    const extPath = context.extensionPath;
    const rootDir = path.join(extPath, 'reminder');
    const resourcePath = path.join(rootDir, 'index.html');
    // const rootDir = path.dirname(resourcePath);

    const strSlideSection = asset.getSlideSection(rootDir);

    let html = fs.readFileSync(resourcePath, 'utf-8');
    html = html.replace("<!-- ${strSlideSection} -->", strSlideSection);

    var hello = JSON.stringify(title);
    html = html.replace("<!-- ${hello} -->", hello);

    // const testDir = path.resolve(__dirname, '../reminder');

    dlg_stefanie = new dialog.WebviewDialog<DialogResult>('webview-kedou', rootDir, html, '');

    const result = await dlg_stefanie.getResult();
    dlg_stefanie = null;

    if (result) {
        // vscode.window.showInformationMessage("Webview dialog result: " + JSON.stringify(result));
    } else {
        // vscode.window.showInformationMessage("The webview dialog was cancelled.");
    }
}

export function deactivate() {
}